migration-engine/core/src/migration/datamodel_differ/models.rs

```diff
@@ -0,0 +1,321 @@
+        self.next.directives.iter().filter(is_regular)
+    }
+
+    /// Regular model directives (`@@`) created in `next`.
+    pub(crate) fn created_regular_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.next_regular_directives().filter(move |next_directive| {
+            self.previous_regular_directives()
+                .find(|previous_directive| directives_match(previous_directive, next_directive))
+                .is_none()
+        })
+    }
+
+    /// Regular model directives (`@@`) deleted in `next`.
+    pub(crate) fn deleted_regular_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.previous_regular_directives().filter(move |previous_directive| {
+            self.next_regular_directives()
+                .find(|next_directive| directives_match(previous_directive, next_directive))
+                .is_none()
+        })
+    }
+
+    /// Iterator over the regular model directives (`@@`) present in both `previous` and `next`.
+    pub(crate) fn regular_directive_pairs(&self) -> impl Iterator<Item = DirectiveDiffer<'_>> {
+        self.previous_regular_directives()
+            .filter_map(move |previous_directive| {
+                self.next_regular_directives()
+                    .find(|next_directive| directives_match(previous_directive, next_directive))
+                    .map(|next_directive| DirectiveDiffer {
+                        previous: previous_directive,
+                        next: next_directive,
+                    })
+            })
+    }
+
+    fn previous_repeated_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.previous.directives.iter().filter(is_repeated)
+    }
+
+    fn next_repeated_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.next.directives.iter().filter(is_repeated)
+    }
+
+    pub(crate) fn created_repeated_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.next_repeated_directives().filter(move |next_directive| {
+            self.previous_repeated_directives()
+                .find(|previous_directive| directives_are_identical(previous_directive, next_directive))
+                .is_none()
+        })
+    }
+
+    pub(crate) fn deleted_repeated_directives(&self) -> impl Iterator<Item = &ast::Directive> {
+        self.previous_repeated_directives().filter(move |previous_directive| {
+            self.next_repeated_directives()
+                .find(|next_directive| directives_are_identical(previous_directive, next_directive))
+                .is_none()
+        })
+    }
+}
+
+fn fields_match(previous: &ast::Field, next: &ast::Field) -> bool {
+    previous.name.name == next.name.name
+}
+
+/// Model directives that can appear multiple times on the same model. Unlike others, they cannot be matched based only on the directive name.
+const REPEATED_MODEL_DIRECTIVES: &[&str] = &["unique", "index"];
```

Comment:
This array means that it would not be possible for a connector to define repeated directives at the moment. I think this is fine and not a problem at the moment. Let's just keep it in mind for the future when that might become a problem.

Reply:
I thought about checking whether a directive is actually repeated instead of hardcoding this list. That means we would update a `@@unique` that is on different fields. I didn't think extremely deeply about it but it could be incorrect to do this if there are concurrent edits (e.g. Alice adds `indexName: "MyIndex"` to an `@@unique`, Bob changes the fields the `@@unique` applies to, and we infer that they both updated the same constraint, which seems wrong).

One nice thing that comes out of having a sort of tree of differ structs is that we could pass configuration into `TopDiffer` and it would circulate down the diffing tree, so this list could be returned from the connector instead of hardcoded.

Reply:
or in addition to the built-in ones